azavea/echo-locator
# Update routing on listing interaction

## Overview

Updates the routing so that a selected apartment listing is the destination, rather than the neighborhood centroid, on hover and on click. Clicking a listing’s marker currently redirects to a listing’s url in a new tab. In order to display the listing route while allowing for the click redirect, this PR displays route and trip info on marker hover with with a 500ms delay or immediately on click for when the user returns to the page. The listings’ routing was built following the same process for neighborhood routing—although simplified since there is only one selected listing and its route represented in state at a time—and successful routing includes: correct route path drawn on map, accurate trip info in Dock, and resetting routing when listing marker is no longer displayed.


### Demo
Route layer should update to apartment listing as destination:
![Screen Shot 2021-12-15 at 11 23 06 AM](https://user-images.githubusercontent.com/36374797/146224851-5e94aaac-6ed7-48f0-a8d1-b9f11e79fa7f.png)
Trip info should update accordingly:
![Screen Shot 2021-12-15 at 11 23 28 AM](https://user-images.githubusercontent.com/36374797/146224912-35a54c2e-e886-4657-bed7-c655f93300ba.png)
If apartment listing is beyond `MAX_TRAVEL_TIME` or the trip is undefined, no route displays and an error message displays in Dock:
![Screen Shot 2021-12-15 at 11 25 45 AM](https://user-images.githubusercontent.com/36374797/146225372-4eb4811d-e6e7-48f1-9560-2dbc377de64c.png)


### Notes

- Since there is a link to directions in the side Dock that users should be able to navigate to without losing the route/trip info, the `mouseout` action of a marker does not reset the route to the centroid of a neighborhood if a user has clicked the listing.
- Users have the option of selecting transit or car for transportation preference. There is no road network that can be used to create a routable "by car" journey and building support for that feature isn't a viable option at the time, so the current solution is to display a "drivable route" represented by a straight line from origin to destination. Similar to transit routing, destination is the neighborhood centroid by default and will update to an apartment listing when selected.


## Testing Instructions

- Run ./scripts/server
- Log in and select a neighborhood and select 'Show BHA' or 'Show Realtor' in the top bar
- Confirm that route updates to route to marker on hover and on click
- Confirm that trip info in Dock updates to route to marker on hover and on click
- Navigate to Haverhill – 01832 neighborhood and select Realtor listing that exceeds `MAX_TRAVEL_TIME` and confirm error message in Dock and no route appears
- On 'Hide BHA' or 'Hide Realtor' confirm that route resets to neighborhood centroid
- Navigate to Edit Profile, toggle between 'car' and 'Bus and/or train' to ensure routing accurately updates by transportation type


Resolves #357
